numba/cuda/simulator/api.py

```diff
@@ -85,9 +85,14 @@ def jit(func_or_sig=None, device=False, debug=False, argtypes=None,
 
     if link is not None:
         raise NotImplementedError('Cannot link PTX in the simulator')
+
+    # Types would imply that the first argument is a signature or sequence of
+    # signatures
+    sig_types = (str, tuple, list, Signature)
```

Comment:
I'm not sure that this matches what's accepted here in the dispatcher? https://github.com/numba/numba/blob/0b1ef84c0d7311e7ff47d3bdfa72c02bb6905b94/numba/cuda/decorators.py#L86-L93

Also, are the unit tests checking that string, `Signature`, list and tuple all work as methods of describing a number of signatures (both for the real dispatcher and the simulator dispatcher)? The new tests seem to cover "list of strings"? Should the accepted "containers" of signatures just be anything that implements the iterator protocol?

Reply:
> I'm not sure that this matches what's accepted here in the dispatcher? 

What do you think the mismatch is? `is_signature()` returns `True` for strings, tuples, and `Signature`s, and the list case is handled in the `elif` clause - am I missing something?

> Also, are the unit tests checking that string, Signature, list and tuple all work as methods of describing a number of signatures (both for the real dispatcher and the simulator dispatcher)? The new tests seem to cover "list of strings"? 

I wrote the new tests by copying and modifying all the relevant ones for the CPU dispatcher, so maybe the testing there is lacking these other cases too.

> Should the accepted "containers" of signatures just be anything that implements the iterator protocol?

I think we generally try to steer people away from eager compilation, so I think I'd not lean towards expanding the API for it without good reason - I think that would be best addressed in a separate PR / discussion if we are going to modify the API.

Reply:
> > I'm not sure that this matches what's accepted here in the dispatcher?
> 
> What do you think the mismatch is? `is_signature()` returns `True` for strings, tuples, and `Signature`s, and the list case is handled in the `elif` clause - am I missing something?

Thanks for pointing this out, in which case, would something like this be a possible improvement in that it doesn't explicitly encode the behaviour of `is_signature` into types?

```python
if func_or_sig is None or sigutils.is_signature(func_or_sig) or isinstance(func_or_sig, list):
```


> > Also, are the unit tests checking that string, Signature, list and tuple all work as methods of describing a number of signatures (both for the real dispatcher and the simulator dispatcher)? The new tests seem to cover "list of strings"?
> 
> I wrote the new tests by copying and modifying all the relevant ones for the CPU dispatcher, so maybe the testing there is lacking these other cases too.

This may indeed be the case. Fixing is probably out of scope for this PR.

> > Should the accepted "containers" of signatures just be anything that implements the iterator protocol?
> 
> I think we generally try to steer people away from eager compilation, so I think I'd not lean towards expanding the API for it without good reason - I think that would be best addressed in a separate PR / discussion if we are going to modify the API.

OK, seems like it's best to leave this for now.

Reply:
> Thanks for pointing this out, in which case, would something like this be a possible improvement in that it doesn't explicitly encode the behaviour of `is_signature` into types?

Given that different things happen in the different branches, I'm not sure I see how that would work... Can you clarify what you're asking please?

Reply:
OK, thanks for OOB clarification - change pushed.

Reply:
From OOB conversation. The guard presently looks a bit like:
```python
if X is None or isinstance(X, <types>)...:
```
where the `<types>` are implicitly encoded by what `is_signature` determines to be a signature, so perhaps this should actually just use `is_signature` so as to remove the implicitness. i.e.
```python
if X is None or sigutils.is_signature(X)...:
```
Thanks for addressing this in acd8e53!